### PR TITLE
Adding Packer Integration

### PR DESCRIPTION
**Pull request by tfhartmann**
This is my active development branch for adding Packer Integration to Leeroy, as of the most recent commit, this actually works! 

Comments welcome for things that should get cleaned, this whole branch needs a rebase, and some tests. But I figured I'd open a PR to get get a code review started! 

.env looks like this 

```
ENVIRONMENT=development
SPEC_OPTS='--format documentation'
LEEROY_PACKER_TEMPLATE_PREFIX=/Users/alaric/git/packer-rk-apps
LEEROY_APP_NAME=rk-bastion
AWS_REGION=us-east-1
LEEROY_AWS_LINUX_AMI=ami-c481fad3
```

Currently the output looks like this!

```Shell
2016-11-16T11:45:44-05:00 [DEBUG] 26283 (initialize): initializing Leeroy::Env
2016-11-16T11:45:44-05:00 [DEBUG] 26283 (initialize): options: {}
2016-11-16T11:45:44-05:00 [DEBUG] 26283 (_filterEnv): filtering env by prefix 'LEEROY_'
2016-11-16T11:45:44-05:00 [DEBUG] 26283 (initialize): filtered: {"LEEROY_AWS_LINUX_AMI"=>"ami-c481fad3", "LEEROY_PACKER_TEMPLATE_PREFIX"=>"/Users/alaric/git/packer-rk-apps", "LEEROY_APP_NAME"=>"rk-bastion"}
2016-11-16T11:45:44-05:00 [DEBUG] 26283 (load_state): loading state from stdin if available
2016-11-16T11:45:44-05:00 [DEBUG] 26283 (rotate_task_metadata): rotating task metadata
2016-11-16T11:45:44-05:00 [DEBUG] 26283 (initialize): initializing AWS helpers
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (initialize): AWS helpers initialized
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (initialize): initializing Packer helpers
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (initialize): Packer helpers initialized
2016-11-16T11:45:46-05:00 [ INFO] 26283 (): performing Leeroy::Task::Packer
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (perform): args: []
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (perform): options: {"p"=>"gold_master", :p=>"gold_master", "phase"=>"gold_master", :phase=>"gold_master", "i"=>nil, :i=>nil, "imageid"=>nil, :imageid=>nil, "n"=>nil, :n=>nil, "name"=>nil, :name=>nil}
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (perform): global_options: {"op"=>true, :op=>true, "stdin"=>true, :stdin=>true, "help"=>false, :help=>false}
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (rescue in fetch): key not found: "phase"
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (rescue in fetch): property 'phase' not found in statedata, checking state
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (perform): phase: gold_master
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (_getPackerParams): generating Packer params to create an AMI
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (checkEnv): checking for 'LEEROY_AWS_LINUX_AMI' in environment
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (checkEnv): candidate: ami-c481fad3
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (checkEnv): check_passed: true
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (checkEnv): checking for 'LEEROY_APP_NAME' in environment
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (checkEnv): candidate: rk-bastion
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (checkEnv): check_passed: true
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (checkEnv): checking for 'LEEROY_PACKER_TEMPLATE_PREFIX' in environment
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (checkEnv): candidate: /Users/alaric/git/packer-rk-apps
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (checkEnv): check_passed: true
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (checkEnv): checking for 'LEEROY_APP_NAME' in environment
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (checkEnv): candidate: rk-bastion
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (checkEnv): check_passed: true
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (validatePacker): Loading Packer using version :'0.12.0'
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (validatePacker): Loading Packer Template :'/Users/alaric/git/packer-rk-apps/rk-bastion/main.json'
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (validatePacker): Validating Packer Template syntax :'/Users/alaric/git/packer-rk-apps/rk-bastion/main.json'
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (buildPacker): Building Packer using version :'0.12.0'
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (buildPacker): Loading Packer Template :'/Users/alaric/git/packer-rk-apps/rk-bastion/main.json'
2016-11-16T11:45:46-05:00 [DEBUG] 26283 (buildPacker): Building Packer Template :'/Users/alaric/git/packer-rk-apps/rk-bastion/main.json'
2016-11-16T11:51:37-05:00 [DEBUG] 26283 (perform): 1479314746,,ui,say,amazon-ebs output will be in this color.
1479314746,,ui,say,
1479314746,,ui,say,==> amazon-ebs: Prevalidating AMI Name...
1479314747,,ui,message,    amazon-ebs: Found Image ID: ami-c481fad3
1479314747,,ui,say,==> amazon-ebs: Creating temporary keypair: packer_582c8d3a-8b70-1eb2-907e-4533e084ccef
1479314747,,ui,say,==> amazon-ebs: Creating temporary security group for this instance...
1479314747,,ui,say,==> amazon-ebs: Authorizing access to port 22 the temporary security group...
1479314748,,ui,say,==> amazon-ebs: Launching a source AWS instance...
1479314748,,ui,message,    amazon-ebs: Instance ID: i-07ec9bbd85e908efc
1479314748,,ui,say,==> amazon-ebs: Waiting for instance (i-07ec9bbd85e908efc) to become ready...
1479314763,,ui,say,==> amazon-ebs: Waiting for SSH to become available...
1479314814,,ui,say,==> amazon-ebs: Connected to SSH!
1479314814,,ui,say,==> amazon-ebs: Uploading ../common/files/_gemrc => /home/ec2-user/_gemrc
1479314814,,ui,say,==> amazon-ebs: Uploading goss.yaml => /home/ec2-user/goss.yaml
1479314814,,ui,say,==> amazon-ebs: Uploading ../common/files/goss => /home/ec2-user/goss
1479314818,,ui,say,==> amazon-ebs: Provisioning with shell script: /var/folders/42/y403wllj2z7_pdqzvcl_4rnm0000gn/T/packer-shell972052736
1479314848,,ui,message,    amazon-ebs: Loaded plugins: priorities%!(PACKER_COMMA) update-motd%!(PACKER_COMMA) upgrade-helper

... packer does a bunch of stuff ...

1479314927,,ui,say,==> amazon-ebs: Stopping the source instance...
1479314927,,ui,say,==> amazon-ebs: Waiting for the instance to stop...
1479314957,,ui,say,==> amazon-ebs: Creating the AMI: rk-bastion-1479314746
1479314957,,ui,message,    amazon-ebs: AMI: ami-1af6dd0d
1479314957,,ui,say,==> amazon-ebs: Waiting for AMI to become ready...
1479315089,,ui,say,==> amazon-ebs: Modifying attributes on AMI (ami-1af6dd0d)...
1479315089,,ui,message,    amazon-ebs: Modifying: description
1479315089,,ui,say,==> amazon-ebs: Terminating the source AWS instance...
1479315096,,ui,say,==> amazon-ebs: Cleaning up any extra volumes...
1479315096,,ui,say,==> amazon-ebs: No volumes to clean up%!(PACKER_COMMA) skipping
1479315096,,ui,say,==> amazon-ebs: Deleting temporary security group...
1479315096,,ui,say,==> amazon-ebs: Deleting temporary keypair...
1479315096,,ui,say,Build 'amazon-ebs' finished.
1479315096,,ui,say,\n==> Builds finished. The artifacts of successful builds are:
1479315096,amazon-ebs,artifact-count,1
1479315096,amazon-ebs,artifact,0,builder-id,mitchellh.amazonebs
1479315096,amazon-ebs,artifact,0,id,us-east-1:ami-1af6dd0d
1479315096,amazon-ebs,artifact,0,string,AMIs were created:\n\nus-east-1: ami-1af6dd0d
1479315096,amazon-ebs,artifact,0,files-count,0
1479315096,amazon-ebs,artifact,0,end
1479315096,,ui,say,--> amazon-ebs: AMIs were created:\n\nus-east-1: ami-1af6dd0d

2016-11-16T11:51:37-05:00 [DEBUG] 26283 (perform): Packer Artifact Created : ami-1af6dd0d
2016-11-16T11:51:37-05:00 [DEBUG] 26283 (dump_state): dumping state to stdout
2016-11-16T11:51:37-05:00 [DEBUG] 26283 (dump): beginning dump
2016-11-16T11:51:37-05:00 [DEBUG] 26283 (dump): dump_properties: [:data, :metadata]
2016-11-16T11:51:37-05:00 [DEBUG] 26283 (block in dump): dumping property 'data'
2016-11-16T11:51:37-05:00 [DEBUG] 26283 (rescue in fetch): key not found: "data"
2016-11-16T11:51:37-05:00 [DEBUG] 26283 (rescue in fetch): property 'data' not found in statedata, checking state
2016-11-16T11:51:37-05:00 [DEBUG] 26283 (block in dump): dumping property 'metadata'
2016-11-16T11:51:37-05:00 [DEBUG] 26283 (rescue in fetch): key not found: "metadata"
2016-11-16T11:51:37-05:00 [DEBUG] 26283 (rescue in fetch): property 'metadata' not found in statedata, checking state
2016-11-16T11:51:37-05:00 [DEBUG] 26283 (dump): dump_mash: #<Leeroy::Types::Mash data={"app_name"=>nil, "imageid"=>"ami-1af6dd0d", "instanceid"=>nil, "message"=>"AMIs were created:\\n\\nus-east-1: ami-1af6dd0d", "phase"=>#<Leeroy::Types::Phase:0x007fea35c58f88 @key=:GOLD_MASTER, @value="gold_master", @ord=0>, "semaphore"=>nil, "sgid"=>nil, "subnetid"=>nil, "vpcid"=>nil} metadata={"task"=>"Leeroy::Task::Packer", "previous"=>nil, "started"=>2016-11-16 11:45:44 -0500}>
{"data":{"app_name":null,"imageid":"ami-1af6dd0d","instanceid":null,"message":"AMIs were created:\\n\\nus-east-1: ami-1af6dd0d","phase":"gold_master","semaphore":null,"sgid":null,"subnetid":null,"vpcid":null},"metadata":{"task":"Leeroy::Task::Packer","previous":null,"started":"2016-11-16 11:45:44 -0500"}}
2016-11-16T11:51:37-05:00 [DEBUG] 26283 (perform): done performing for Leeroy::Task::Packer
```

### CODE REVIEW

File: lib/leeroy/task/packer.rb [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+            app_name = checkEnv('LEEROY_APP_NAME')
+          else
+            app_name = options[:name]
+          end
+          packer_params.app_name = app_name
+
+          if self.state.aws_region?
+            aws_region = self.state.aws_region
+          else
+            aws_region = ENV['AWS_REGION']
```
Comment (tfhartmann) [inline lib/leeroy/task/packer.rb]: This doesn't feel like the best way to do this, but checkEnv doesn't pick up AWS_REGION at the moment.

Comment (hakamadare) [inline lib/leeroy/task/packer.rb]: i think `AWS_DEFAULT_REGION` is actually the env var that the AWS libraries use; we should probably stop using `AWS_REGION`

Comment (tfhartmann) [inline lib/leeroy/task/packer.rb]: I could do this `aws_region = ENV['AWS_DEFAULT_REGION'] || ENV['AWS_REGION']` which would work as a bridge for until, unless we deprecate `AWS_REGION`  That was if it's been set for the aws-cli with `AWS_REGION` the packer task will still work as expected. 

Just tested it locally by unsetting `AWS_REGION` and setting `AWS_DEFAULT_REGION` and it works!

**Comment by tfhartmann** [comment]
So, `leeroy packer -p gold_master | leeroy instantiate -p application | leeroy terminate` now does the thing that I expect it to!

File: lib/leeroy/task/packer.rb [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+          self.state.app_name = packer_vars.app_name
+
+
+          cwd = File.join(checkEnv('LEEROY_PACKER_TEMPLATE_PREFIX'), checkEnv('LEEROY_APP_NAME'))
+
+          validation = validatePacker(cwd, { :vars => packer_vars })
+
+          build = buildPacker(cwd,{ :vars => packer_vars } )
+          build.artifacts.each do | item |
```
Comment (tfhartmann) [inline lib/leeroy/task/packer.rb]: Right now the object we get back looks like this, i'll try a test and see what we get back when you create an ami in multiple regions...

```Ruby
>> pp build.artifacts
[#<Packer::Message::Artifact:0x007fe62741d6f0
  @artifact_index="0",
  @builder_id="mitchellh.amazonebs",
  @id="us-east-1:ami-1cd1e00b",
  @string="AMIs were created:\\n\\nus-east-1: ami-1cd1e00b",
  @target="amazon-ebs",
  @timestamp="1479843763",
  @type="artifact">]
=> [#<Packer::Message::Artifact:0x007fe62741d6f0 @timestamp="1479843763", @target="amazon-ebs", @type="artifact", @artifact_index="0", @builder_id="mitchellh.amazonebs", @id="us-east-1:ami-1cd1e00b", @string="AMIs were created:\\n\\nus-east-1: ami-1cd1e00b">]
>> build.artifacts.class
=> Array
>>
```

Comment (hakamadare) [inline lib/leeroy/task/packer.rb]: a bit alarming that you could potentially have a `Packer::Message::Artifact` whose `type` is something other than "artifact" :trollface:

File: spec/leeroy_spec.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    @env = Leeroy::Env.new
+  end
+  it "get instance of Leeroy::Env" do
+    expect(@env).to be_instance_of(Leeroy::Env)
+  end
+end
+
+describe "Test Leeroy::Task::Fixture New" do
+  before(:each) do
+    @fixture = Leeroy::Task::Fixture.new
```
Comment (hakamadare) [inline spec/leeroy_spec.rb]: also: this test conversation is potentially longer, i don't want it to block the merge of this PR

**Comment by hakamadare** [comment]
:shipit: 💯

File: spec/leeroy_spec.rb (hunk 3 above)
Comment (tfhartmann) [inline spec/leeroy_spec.rb]: Yeah, and lets see if we can ~~sucker~~ , er convince @dcrelling to drop some unit testing words of wisdom on us too 😃

File: lib/leeroy/task/packer.rb (hunk 2 above)
Comment (tfhartmann) [inline lib/leeroy/task/packer.rb]: before I merge, I'm going to check the result of the above output when I copy it to us-east-1 and us-east-2, running that now :)

Comment (tfhartmann) [inline lib/leeroy/task/packer.rb]: Well I got my answer and it made me say :doh:  I think we'll just need to parse this string correctly 

`id="us-east-1:ami-40023957%!(PACKER_COMMA)us-east-2:ami-258cd640",`

Full ouptut...

```Ruby
[#<Packer::Message::Artifact:0x007fd770802c68
  @artifact_index="0",
  @builder_id="mitchellh.amazonebs",
  @id="us-east-1:ami-40023957%!(PACKER_COMMA)us-east-2:ami-258cd640",
  @string=
   "AMIs were created:\\n\\nus-east-1: ami-40023957\\nus-east-2: ami-258cd640",
  @target="amazon-ebs",
  @timestamp="1480695261",
  @type="artifact">]
=> [#<Packer::Message::Artifact:0x007fd770802c68 @timestamp="1480695261", @target="amazon-ebs", @type="artifact", @artifact_index="0", @builder_id="mitchellh.amazonebs", @id="us-east-1:ami-40023957%!(PACKER_COMMA)us-east-2:ami-258cd640", @string="AMIs were created:\\n\\nus-east-1: ami-40023957\\nus-east-2: ami-258cd640">]
```

Comment (tfhartmann) [inline lib/leeroy/task/packer.rb]: Ok, after a @hakamadare @tfhartmann mind meld, I'm going to create an issue in the backlog and merge this.   We can tackle this by parsing out all the %!(...) Macros, then pushing region => amiid into a hash

**Pull request merged**